src/domain/services/jwt.subscription.service.ts

```diff
@@ -0,0 +1,133 @@
+import Dataloader from 'dataloader';
+import { Exception } from 'src/domain/entities/exception.entity.js';
+import { JwtProvider } from '../ports/secondary/jwt.provider.js';
+import { SubscriptionRepository } from 'src/domain/ports/secondary/subscription.repository.js';
+import { Frequency } from '../entities/subscription.entity.js';
+import { EmailProvider } from 'src/domain/ports/secondary/email.provider.js';
+import { TemplateProvider } from '../ports/secondary/templates.provider.js';
+import { Logger, LoggerProvider } from '../ports/secondary/logger.provider.js';
+import { Weather } from '../entities/weather.entity.js';
+import { WeatherService } from '../ports/primary/weather.service.js';
+import { SubscriptionService } from '../ports/primary/subscription.service.js';
+
+export type SubscribeOptions = {
+  email: string;
+  city: string;
+  frequency: Frequency;
+};
+
+type Options = {
+  jwtProvider: JwtProvider;
+  subscriptionRepository: SubscriptionRepository;
+  weatherService: WeatherService;
+  emailProvider: EmailProvider;
+  templateProvider: TemplateProvider;
+  loggerProvider: LoggerProvider;
+  config: { APP_URL: string };
+};
+
+export class JwtSubscriptionService implements SubscriptionService {
+  private jwtProvider: JwtProvider;
+  private subscriptionRepository: SubscriptionRepository;
+  private weatherService: WeatherService;
+  private emailProvider: EmailProvider;
+  private templateProvider: TemplateProvider;
+  private logger: Logger;
+
+  private appUrl: string;
+
+  constructor({
+    jwtProvider,
+    subscriptionRepository,
+    weatherService,
+    emailProvider,
+    templateProvider,
+    loggerProvider,
+    config,
+  }: Options) {
+    this.jwtProvider = jwtProvider;
+    this.subscriptionRepository = subscriptionRepository;
+    this.weatherService = weatherService;
+    this.emailProvider = emailProvider;
+    this.templateProvider = templateProvider;
+
+    this.logger = loggerProvider.createLogger('SubscriptionService');
+
+    this.appUrl = config.APP_URL;
+  }
+
+  private makeUnsubscribeLink(subscriptionId: string) {
+    return `${this.appUrl}/api/unsubscribe/${subscriptionId}`;
+  }
+
+  public async handleSubscriptions(frequency: Frequency) {
+    this.logger.info({ msg: 'Handling weather subscription has been started', frequency });
+
+    const dataloader = new Dataloader<string, Weather>(async (cities) => {
+      return await Promise.all(
+        cities.map(async (city) => {
+          return await this.weatherService.getWeather(city);
+        }),
+      );
+    });
```

Comment:
It would be great to shorten it to something like this to simplify reading
```suggestion
    const dataloader = new Dataloader<string, Weather>(
      (cities) => Promise.all(
        cities.map(this.weatherService.getWeather),
      );
    );
```

Reply:
But then we will have wrong stacktraces:
https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/errorhandling/returningpromises.md
https://t.me/node_recipes/590

Reply:
Then keep one await not 2